xray/aws_v2.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2017-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+package xray
```

Comment:
Can you move this into another package? We found the current pattern bloats binary size so we should aim for a package for each instrumentation

Reply:
I have added the separate module for this PR but this change would require a lot of re working because currently we release the repository as a single package which includes instrumentation packages and SDK. Now, If we want to create a separate module for each instrumentation package then I think that would require moving things around and basically separate out the SDK and instrumentation packages. Once we do that then, we can manage separate release of core SDK and instrumentation packages. I am not able to visualize the release process in this scenario where we add separate instrumentation package at the root of this repository.

Reply:
I was hoping having just a separate package, without a separate go.mod, would be enough for go to not include unneeded code in a downstream binary. Do you know if this is the case or not?

Reply:
Actually thinking more don't see a problem with the current PR adding a go.mod. What do you mean about the release process? Isn't it just cutting a tag in this repo?

We can't really move around the existing instrumentation without a major version bump so we'll hold off on that, but we may as well try to split better for new instrumentation.

Reply:
Basically, currently we are tagging the entire repository as released version. If we separate instrumentation package then It means some modules in the instrumentation packages relies on core `xray` package. So, I am not able to visualize how we are going to tag these packages with out reworking the SDK. Ideally what we should do is to have separate module for core SDK and separate module for instrumentation and we should follow the release cycle of core SDK and separate instrumentation packages rather than tagging an entire repository.  If we take open telemetry contrib repo as an example here's how it released packages looks like: https://github.com/open-telemetry/opentelemetry-go-contrib/releases . I am not sure about this - https://github.com/aws/aws-xray-sdk-go/pull/309#discussion_r653293298 AFAIK In order to reduce the dependencies it has to be the module but I will dig more and update the thread.

Reply:
Any concern if we just create separate go.mod for aws sdk v2 instrumentation, not change legacy instrumentations?

Reply:
@wangzlei As I have mentioned it would require moving things around within SDK. Right now instrumentation packages and core SDK packages resides in `xray` dir. We probably have to separate out those components. @anuraaga and I had a discussion regarding the same so we end up deciding to have a separate package for aws v2 instrumentation and not a module right now to gain some advantage of dead code elimination. Also, this would require a major version release since this is a breaking change.